File: scripts/test_connection.py
#!/usr/bin/env python3
"""Test script to verify Toggl Track API connection."""

import asyncio
import os
import sys
from datetime import datetime, timedelta

from dotenv import load_dotenv

# Add the parent directory to the path so we can import our modules
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from toggl_track_mcp.toggl_client import TogglAPIClient, TogglAPIError


async def test_connection():
    """Test Toggl Track API connection and functionality."""
    load_dotenv()
    
    api_token = os.getenv("TOGGL_API_TOKEN")
    if not api_token:
        print("❌ Error: TOGGL_API_TOKEN environment variable not set")
        print("Please set your Toggl Track API token in the .env file")
        return False
    
    print("🔍 Testing Toggl Track API connection...\n")
    
    try:
        # Initialize client
        client = TogglAPIClient(api_token=api_token)
        
        # Test 1: Get current user
        print("👤 Testing user authentication...")
        user = await client.get_current_user()
        print(f"✅ Authenticated as: {user.fullname} ({user.email})")
        print(f"   Default workspace: {user.default_workspace_id}")
        print(f"   Timezone: {user.timezone}\n")
        
        # Test 2: Get workspaces
        print("🏢 Testing workspace access...")
        workspaces = await client.get_workspaces()
        print(f"✅ Found {len(workspaces)} workspace(s):")
        for ws in workspaces:
            print(f"   - {ws.name} (ID: {ws.id}, Admin: {ws.admin})")
        print()
        
        # Test 3: Get projects
        print("📂 Testing project access...")
        projects = await client.get_projects()
        print(f"✅ Found {len(projects)} project(s):")
        active_projects = [p for p in projects if p.active]
        for project in active_projects[:5]:  # Show first 5 active projects
            billable_status = "billable" if project.billable else "non-billable"
            print(f"   - {project.name} ({billable_status})")
        if len(active_projects) > 5:
            print(f"   ... and {len(active_projects) - 5} more active projects")
        print()
        
        # Test 4: Get clients
        print("👥 Testing client access...")
        clients = await client.get_clients()
        print(f"✅ Found {len(clients)} client(s):")
        for client_obj in clients[:5]:  # Show first 5 clients
            print(f"   - {client_obj.name}")
        if len(clients) > 5:
            print(f"   ... and {len(clients) - 5} more clients")
        print()
        
        # Test 5: Get current time entry
        print("⏱️  Testing current time entry...")
        current_entry = await client.get_current_time_entry()
        if current_entry:
            duration = client.calculate_duration(current_entry)
            duration_formatted = client.format_duration(duration)
            print(f"✅ Timer is running: '{current_entry.description}'")
            print(f"   Duration: {duration_formatted}")
            print(f"   Project ID: {current_entry.project_id or 'No project'}")
            print(f"   Billable: {current_entry.billable}")
        else:
            print("⏸️  No timer currently running")
        print()
        
        # Test 6: Get recent time entries
        print("📝 Testing time entries access...")
        # Get entries from last 7 days
        end_date = datetime.now()
        start_date = end_date - timedelta(days=7)
        
        entries = await client.get_time_entries(
            start_date=start_date.strftime("%Y-%m-%d"),
            end_date=end_date.strftime("%Y-%m-%d")
        )
        print(f"✅ Found {len(entries)} time entries in the last 7 days:")
        
        total_duration = 0
        billable_duration = 0
        for entry in entries[:3]:  # Show first 3 entries
            duration = client.calculate_duration(entry)
            duration_formatted = client.format_duration(duration)
            total_duration += duration
            if entry.billable:
                billable_duration += duration
            
            status = "running" if entry.duration < 0 else "completed"
            billable_status = "billable" if entry.billable else "non-billable"
            print(f"   - '{entry.description}' ({duration_formatted}, {billable_status}, {status})")
        
        if len(entries) > 3:
            print(f"   ... and {len(entries) - 3} more entries")
        
        if entries:
            total_formatted = client.format_duration(total_duration)
            billable_formatted = client.format_duration(billable_duration)
            print(f"   Total time: {total_formatted} ({billable_formatted} billable)")
        print()
        
        # Test 7: Get tags
        print("🏷️  Testing tags access...")
        tags = await client.get_tags()
        print(f"✅ Found {len(tags)} tag(s):")
        for tag in tags[:10]:  # Show first 10 tags
            print(f"   - {tag.name}")
        if len(tags) > 10:
            print(f"   ... and {len(tags) - 10} more tags")
        print()
        
        # Test 8: Rate limiting
        print("🚦 Testing rate limiting...")
        start_time = datetime.now()
        for i in range(3):
            await client.get_current_user()
        end_time = datetime.now()
        duration = (end_time - start_time).total_seconds()
        print(f"✅ Made 3 requests in {duration:.2f} seconds (rate limiting working)")
        print()
        
        print("🎉 All tests passed! Toggl Track MCP server should work correctly.")
        print("\nNext steps:")
        print("1. Install dependencies: pip install -e .")
        print("2. Run the server: python -m uvicorn toggl_track_mcp.server:app --reload")
        print("3. Test MCP endpoints at http://localhost:8000/mcp")
        
        return True
        
    except TogglAPIError as e:
        print(f"❌ Toggl API Error: {e}")
        if e.status_code == 401:
            print("\n💡 Authentication failed. Please check:")
            print("   - Your TOGGL_API_TOKEN in the .env file")
            print("   - That the token is copied correctly from your Toggl profile")
            print("   - Go to https://track.toggl.com/profile to get your API token")
        elif e.status_code == 403:
            print("\n💡 Permission denied. Please check:")
            print("   - Your Toggl account has the necessary permissions")
            print("   - You have access to the workspace")
        elif e.status_code == 402:
            print("\n💡 Payment required. Please check:")
            print("   - Your Toggl subscription status")
            print("   - API access may require a paid plan")
        return False
        
    except Exception as e:
        print(f"❌ Unexpected error: {e}")
        print("\n💡 This might be a network issue or configuration problem.")
        print("Please check:")
        print("   - Your internet connection")
        print("   - That you can access https://api.track.toggl.com")
        print("   - Your .env file configuration")
        return False


if __name__ == "__main__":
    success = asyncio.run(test_connection())
    sys.exit(0 if success else 1)

File: scripts/test-connection.ts
#!/usr/bin/env tsx

import 'dotenv/config';
import { TogglClient } from '../src/toggl-client.js';
import { loadConfig } from '../src/config.js';

async function testConnection(): Promise<void> {
  try {
    console.log('Testing Toggl Track API connection...\n');
    
    const config = loadConfig();
    const client = new TogglClient(config);
    
    console.log('🔍 Fetching current user...');
    const user = await client.getCurrentUser();
    console.log(`✅ User: ${user.fullname} (${user.email})`);
    console.log(`   Default Workspace: ${user.default_workspace_id}`);
    console.log(`   Timezone: ${user.timezone}\n`);
    
    console.log('🏢 Fetching workspaces...');
    const workspaces = await client.getWorkspaces();
    console.log(`✅ Found ${workspaces.length} workspace(s):`);
    workspaces.forEach(ws => {
      console.log(`   - ${ws.name} (ID: ${ws.id})`);
    });
    console.log();
    
    console.log('📂 Fetching projects...');
    const projects = await client.getProjects();
    console.log(`✅ Found ${projects.length} project(s):`);
    projects.slice(0, 5).forEach(project => {
      console.log(`   - ${project.name} (ID: ${project.id}, Active: ${project.active})`);
    });
    if (projects.length > 5) {
      console.log(`   ... and ${projects.length - 5} more`);
    }
    console.log();
    
    console.log('👥 Fetching clients...');
    const clients = await client.getClients();
    console.log(`✅ Found ${clients.length} client(s):`);
    clients.slice(0, 5).forEach(client => {
      console.log(`   - ${client.name} (ID: ${client.id})`);
    });
    if (clients.length > 5) {
      console.log(`   ... and ${clients.length - 5} more`);
    }
    console.log();
    
    console.log('🏃 Checking for current time entry...');
    const currentEntry = await client.getCurrentTimeEntry();
    if (currentEntry) {
      const duration = currentEntry.duration < 0 
        ? Math.abs(Date.now() / 1000 + currentEntry.duration)
        : currentEntry.duration;
      console.log(`✅ Timer is running: "${currentEntry.description}"`);
      console.log(`   Duration: ${Math.floor(duration / 3600)}h ${Math.floor((duration % 3600) / 60)}m`);
      console.log(`   Project: ${currentEntry.project_id || 'No project'}`);
    } else {
      console.log('⏸️  No timer currently running');
    }
    console.log();
    
    console.log('📝 Fetching recent time entries...');
    const entries = await client.getTimeEntries({
      startDate: new Date(Date.now() - 7 * 24 * 60 * 60 * 1000).toISOString().split('T')[0]
    });
    console.log(`✅ Found ${entries.length} time entries in the last 7 days`);
    entries.slice(0, 3).forEach(entry => {
      const duration = entry.duration < 0 
        ? Math.abs(Date.now() / 1000 + entry.duration)
        : entry.duration;
      const hours = Math.floor(duration / 3600);
      const minutes = Math.floor((duration % 3600) / 60);
      console.log(`   - "${entry.description}" (${hours}h ${minutes}m, ${entry.billable ? 'billable' : 'non-billable'})`);
    });
    console.log();
    
    console.log('🎯 Fetching tags...');
    const tags = await client.getTags();
    console.log(`✅ Found ${tags.length} tag(s):`);
    tags.slice(0, 10).forEach(tag => {
      console.log(`   - ${tag.name}`);
    });
    if (tags.length > 10) {
      console.log(`   ... and ${tags.length - 10} more`);
    }
    console.log();
    
    console.log('🎉 Connection test completed successfully!');
    console.log('\nThe Toggl Track MCP server should work correctly with your API token.');
    
  } catch (error) {
    console.error('❌ Connection test failed:');
    console.error(error instanceof Error ? error.message : String(error));
    
    if (error instanceof Error) {
      if (error.message.includes('401') || error.message.includes('Unauthorized')) {
        console.error('\n💡 This looks like an authentication error. Please check:');
        console.error('   - Your TOGGL_API_TOKEN in the .env file');
        console.error('   - That the token is copied correctly from your Toggl profile');
      } else if (error.message.includes('403') || error.message.includes('Forbidden')) {
        console.error('\n💡 This looks like a permissions error. Please check:');
        console.error('   - That your Toggl account has API access');
        console.error('   - Your workspace permissions');
      } else if (error.message.includes('network') || error.message.includes('ENOTFOUND')) {
        console.error('\n💡 This looks like a network error. Please check:');
        console.error('   - Your internet connection');
        console.error('   - That you can access https://api.track.toggl.com');
      }
    }
    
    process.exit(1);
  }
}

testConnection();